File: packages/core/src/explain-tamper.ts
import { hashCanonical, hashPayload } from "./hashing.js";
import { GENESIS_HASH } from "./chain.js";
import type { PassportRecord, ChainManifest } from "./types.js";

export interface TamperFinding {
  recordIndex: number;
  recordId: string;
  kind: "payload_hash" | "record_hash" | "prev_hash" | "sequence" | "manifest_chain_hash";
  expected: string;
  actual: string;
  message: string;
}

export interface TamperExplanation {
  tampered: boolean;
  findings: TamperFinding[];
  summary: string;
}

export function explainTamper(
  records: PassportRecord[],
  manifest?: ChainManifest,
): TamperExplanation {
  const findings: TamperFinding[] = [];

  for (let i = 0; i < records.length; i++) {
    const record = records[i];

    // Check sequence
    if (record.sequence !== i) {
      findings.push({
        recordIndex: i,
        recordId: record.id,
        kind: "sequence",
        expected: String(i),
        actual: String(record.sequence),
        message: `Record ${i} has sequence ${record.sequence}, expected ${i}`,
      });
    }

    // Check prev_hash linkage
    const expectedPrev = i === 0 ? GENESIS_HASH : records[i - 1].record_hash;
    if (record.prev_hash !== expectedPrev) {
      findings.push({
        recordIndex: i,
        recordId: record.id,
        kind: "prev_hash",
        expected: expectedPrev,
        actual: record.prev_hash,
        message: `Record ${i} prev_hash does not match record ${i - 1} hash — chain link broken`,
      });
    }

    // Check payload_hash
    const recomputedPayloadHash = hashPayload(record.payload);
    if (recomputedPayloadHash !== record.payload_hash) {
      findings.push({
        recordIndex: i,
        recordId: record.id,
        kind: "payload_hash",
        expected: recomputedPayloadHash,
        actual: record.payload_hash,
        message: `Record ${i} payload was modified — payload hash mismatch`,
      });
    }

    // Check record_hash
    const { record_hash, ...rest } = record;
    const recomputedRecordHash = hashCanonical(rest);
    if (recomputedRecordHash !== record_hash) {
      findings.push({
        recordIndex: i,
        recordId: record.id,
        kind: "record_hash",
        expected: recomputedRecordHash,
        actual: record_hash,
        message: `Record ${i} record_hash mismatch — record content was altered`,
      });
    }
  }

  // Check manifest chain_hash if provided
  if (manifest && records.length > 0) {
    const lastHash = records[records.length - 1].record_hash;
    if (manifest.chain_hash !== lastHash) {
      findings.push({
        recordIndex: records.length - 1,
        recordId: records[records.length - 1].id,
        kind: "manifest_chain_hash",
        expected: lastHash,
        actual: manifest.chain_hash,
        message: `Manifest chain_hash does not match last record hash`,
      });
    }
  }

  const tampered = findings.length > 0;
  let summary: string;

  if (!tampered) {
    summary = "No tampering detected. All records, hashes, and chain links are intact.";
  } else {
    const kinds = [...new Set(findings.map((f) => f.kind))];
    const parts: string[] = [];
    if (kinds.includes("payload_hash")) parts.push("payload content was modified");
    if (kinds.includes("record_hash")) parts.push("record hashes are inconsistent");
    if (kinds.includes("prev_hash")) parts.push("chain links are broken");
    if (kinds.includes("sequence")) parts.push("record sequencing is wrong");
    if (kinds.includes("manifest_chain_hash")) parts.push("manifest does not match chain");
    summary = `Tampering detected in ${findings.length} check(s): ${parts.join("; ")}.`;
  }

  return { tampered, findings, summary };
}

File: python/decision_passport_py/src/decision_passport/explain_tamper.py
from __future__ import annotations

from .chain import GENESIS_HASH
from .hashing import hash_canonical, hash_payload
from .types import ChainManifest, PassportRecord, TamperExplanation, TamperFinding


def explain_tamper_chain(
    records: list[PassportRecord],
    manifest: ChainManifest | None = None,
) -> TamperExplanation:
    findings: list[TamperFinding] = []

    for index, record in enumerate(records):
        if record["sequence"] != index:
            findings.append(
                {
                    "recordIndex": index,
                    "recordId": record["id"],
                    "kind": "sequence",
                    "expected": str(index),
                    "actual": str(record["sequence"]),
                    "message": f"Record {index} has sequence {record['sequence']}, expected {index}",
                }
            )

        expected_prev = GENESIS_HASH if index == 0 else records[index - 1]["record_hash"]
        if record["prev_hash"] != expected_prev:
            findings.append(
                {
                    "recordIndex": index,
                    "recordId": record["id"],
                    "kind": "prev_hash",
                    "expected": expected_prev,
                    "actual": record["prev_hash"],
                    "message": f"Record {index} prev_hash does not match record {index - 1} hash — chain link broken",
                }
            )

        recomputed_payload = hash_payload(record["payload"])
        if recomputed_payload != record["payload_hash"]:
            findings.append(
                {
                    "recordIndex": index,
                    "recordId": record["id"],
                    "kind": "payload_hash",
                    "expected": recomputed_payload,
                    "actual": record["payload_hash"],
                    "message": f"Record {index} payload was modified — payload hash mismatch",
                }
            )

        stored_record_hash = record["record_hash"]
        rest = {k: v for k, v in record.items() if k != "record_hash"}
        recomputed_record_hash = hash_canonical(rest)
        if recomputed_record_hash != stored_record_hash:
            findings.append(
                {
                    "recordIndex": index,
                    "recordId": record["id"],
                    "kind": "record_hash",
                    "expected": recomputed_record_hash,
                    "actual": stored_record_hash,
                    "message": f"Record {index} record_hash mismatch — record content was altered",
                }
            )

    if manifest is not None and len(records) > 0:
        last_hash = records[-1]["record_hash"]
        if manifest["chain_hash"] != last_hash:
            findings.append(
                {
                    "recordIndex": len(records) - 1,
                    "recordId": records[-1]["id"],
                    "kind": "manifest_chain_hash",
                    "expected": last_hash,
                    "actual": manifest["chain_hash"],
                    "message": "Manifest chain_hash does not match last record hash",
                }
            )

    tampered = len(findings) > 0
    if not tampered:
        summary = "No tampering detected. All records, hashes, and chain links are intact."
    else:
        kinds = {f["kind"] for f in findings}
        parts: list[str] = []
        if "payload_hash" in kinds:
            parts.append("payload content was modified")
        if "record_hash" in kinds:
            parts.append("record hashes are inconsistent")
        if "prev_hash" in kinds:
            parts.append("chain links are broken")
        if "sequence" in kinds:
            parts.append("record sequencing is wrong")
        if "manifest_chain_hash" in kinds:
            parts.append("manifest does not match chain")
        summary = f"Tampering detected in {len(findings)} check(s): {'; '.join(parts)}."

    return {
        "tampered": tampered,
        "findings": findings,
        "summary": summary,
    }
